transport/awssqs/consumer.go

```diff
@@ -0,0 +1,296 @@
+func ConsumerErrorHandler(errorHandler transport.ErrorHandler) ConsumerOption {
+	return func(c *Consumer) { c.errorHandler = errorHandler }
+}
+
+// ConsumerFinalizer is executed once all the received SQS messages are done being processed.
+// By default, no finalizer is registered.
+func ConsumerFinalizer(f ...ConsumerFinalizerFunc) ConsumerOption {
+	return func(c *Consumer) { c.finalizer = f }
+}
+
+// ConsumerDeleteMessage overrides the default value for the consumer's
+// deleteMessage field to indicate when the consumed messages should be deleted.
+func ConsumerDeleteMessage(delete Delete) ConsumerOption {
+	return func(c *Consumer) { c.deleteMessage = delete }
+}
+
+// Consume calls ReceiveMessageWithContext and handles messages having an
+// sqs.ReceiveMessageInput as parameter allows each user to have his own receive configuration.
+// That said, this method overrides the queueURL for the provided ReceiveMessageInput to ensure
+// the messages are retrieved from the consumer's configured queue.
+func (c Consumer) Consume(ctx context.Context, receiveMsgInput *sqs.ReceiveMessageInput) error {
+	receiveMsgInput.QueueUrl = &c.queueURL
+	out, err := c.sqsClient.ReceiveMessageWithContext(ctx, receiveMsgInput)
+	if err != nil {
+		return err
+	}
+	return c.handleMessages(ctx, out.Messages)
+}
+
+// handleMessages handles the consumed messages.
+func (c Consumer) handleMessages(ctx context.Context, msgs []*sqs.Message) error {
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	// Copy received messages slice in leftMsgs slice
+	// leftMsgs will be used by the consumer's visibilityTimeoutFunc to extend the
+	// visibility timeout for the messages that have not been processed yet.
+	leftMsgs := []*sqs.Message{}
+	leftMsgs = append(leftMsgs, msgs...)
+
+	visibilityTimeoutCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go c.visibilityTimeoutFunc(visibilityTimeoutCtx, c.sqsClient, c.queueURL, c.visibilityTimeout, &leftMsgs, c.leftMsgsMux)
+
+	if len(c.finalizer) > 0 {
+		defer func() {
+			for _, f := range c.finalizer {
+				f(ctx, &msgs)
+			}
+		}()
+	}
+
+	for _, f := range c.before {
+		ctx = f(ctx, &msgs)
+	}
+
+	for _, msg := range msgs {
+		if c.deleteMessage == BeforeHandle {
+			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
+				QueueUrl:      &c.queueURL,
+				ReceiptHandle: msg.ReceiptHandle,
+			}); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				return err
+			}
+		}
+
+		if err := c.handleSingleMessage(ctx, msg, &leftMsgs); err != nil {
+			return err
+		}
+
+		if c.deleteMessage == AfterHandle {
+			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
+				QueueUrl:      &c.queueURL,
+				ReceiptHandle: msg.ReceiptHandle,
+			}); err != nil {
+				c.errorHandler.Handle(ctx, err)
+				c.errorEncoder(ctx, err, msg, c.sqsClient)
+				return err
+			}
+		}
```

Comment:
```suggestion
	for _, f := range c.before {
		ctx = f(ctx, &msgs, c.sqsClient)
	}

	for _, msg := range msgs {
		if c.deleteMessage == BeforeHandle {
			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
				QueueUrl:      &c.queueURL,
				ReceiptHandle: msg.ReceiptHandle,
			}); err != nil {
				c.errorHandler.Handle(ctx, err)
				c.errorEncoder(ctx, err, msg, c.sqsClient)
				return err
			}
		}

		if err := c.handleSingleMessage(ctx, msg, &leftMsgs); err != nil {
			return err
		}

		if c.deleteMessage == AfterHandle {
			if _, err := c.sqsClient.DeleteMessageWithContext(ctx, &sqs.DeleteMessageInput{
				QueueUrl:      &c.queueURL,
				ReceiptHandle: msg.ReceiptHandle,
			}); err != nil {
				c.errorHandler.Handle(ctx, err)
				c.errorEncoder(ctx, err, msg, c.sqsClient)
				return err
			}
		}
```

It can be done on before or after func, right?

I prefer this as plugin if the developer want to delete the message before or after the message is handled.
For sure before function need to append signature which the SQS API.